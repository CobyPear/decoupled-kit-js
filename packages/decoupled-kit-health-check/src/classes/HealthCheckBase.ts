import { logger } from '../utils/logger';
import {
	DecoupledMenuError,
	DecoupledRouterError,
	InvalidCMSEndpointError,
} from './errors';

export abstract class HealthCheckBase {
	abstract endpoint: string;
	abstract envVar: string;
	abstract log: typeof logger;
	/**
	 * Validate the set endpoint
	 */
	abstract validateEndpoint(): Promise<this | InvalidCMSEndpointError>;
	/**
	 * Uses `this.endpoint` to return a new URL
	 */
	abstract getURL(): URL;
	/**
	 * Check the url for a 200 response
	 * @param url - a URL with defined pathname and/or search params
	 */
	abstract checkFor200(url: URL): Promise<boolean>;
	/**
	 * Validate the menu query or endpoint
	 */
	abstract validateMenu(): Promise<this | DecoupledMenuError>;
	/**
	 * Validate the provided credentials
	 */
	abstract validateAuth(): Promise<this>;
	/**
	 * Validate the preview secret is set and preview is configured at the endpoint by fetching preview content.
	 * Should be skipped if credentials are not validated in `this.validateAuth()`.
	 */
	abstract validatePreview(): Promise<this>;
}
export abstract class DrupalHealthCheck extends HealthCheckBase {
	/**
	 * True if the Drupal instance has umami demo data
	 */
	abstract hasUmami: boolean;
	/**
	 * The clientID of the Drupal OAuth client
	 */
	abstract clientID: string | undefined;
	/**
	 * Sets `this.hasUmami` by checking the language settings endpoint
	 */
	abstract checkForUmami(): Promise<void>;
	/**
	 * Validates that the Decoupled Router module is enabled and working
	 */
	abstract validateRouter(): Promise<this | DecoupledRouterError>;
}
export abstract class WordPressHealthCheck extends HealthCheckBase {
	abstract appUsername: string | undefined;
}
