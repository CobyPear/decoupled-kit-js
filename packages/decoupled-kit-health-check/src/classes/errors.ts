class HealthCheckError extends Error {
	constructor(message: string) {
		super(`|__‚ùå ${message}`);
		// no need to show the stack for these errors
		this.stack = undefined;
	}
}

export class BackendNotSetError extends HealthCheckError {
	constructor(
		{ envVar }: { envVar: string },
		message = `No backend set: 
	The PANTHEON_CMS_ENDPOINT or ${envVar} environment variable must be set to fetch data.`,
	) {
		super(message);
	}
}

export class InvalidCMSEndpointError extends HealthCheckError {
	constructor(
		{ endpointType, endpoint }: { endpointType: string; endpoint: string },
		message = `${endpointType} could not be fetched from.
Check that ${endpoint} is valid and provides a 200 response`,
	) {
		super(message);
	}
}

export class DecoupledRouterError extends HealthCheckError {
	constructor(
		{ endpointType, endpoint }: { endpointType: string; endpoint: string },
		message = `Decoupled Router not detected for ${endpointType}.
Check that ${endpoint} is valid and provides a 200 response.
Also ensure that the Decoupled Router module is enabled.`,
	) {
		super(message);
	}
}

export class DecoupledMenuError extends HealthCheckError {
	constructor(
		{ endpointType, endpoint }: { endpointType: string; endpoint: string },
		message = `Decoupled Menu Endpoint not valid for ${endpointType}.
Check that ${endpoint} is valid and provides a 200 response.
Also ensure that the JSON:API Menu Items module is enabled.`,
	) {
		super(message);
	}
}
