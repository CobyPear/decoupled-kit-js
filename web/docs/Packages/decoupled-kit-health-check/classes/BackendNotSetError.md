---
id: "BackendNotSetError"
title: "Class: BackendNotSetError"
sidebar_label: "BackendNotSetError"
sidebar_position: 0
custom_edit_url: null
---

## Hierarchy

- `HealthCheckError`

  ↳ **`BackendNotSetError`**

## Constructors

### constructor

• **new BackendNotSetError**(`«destructured»`, `message?`)

#### Parameters

| Name | Type |
| :------ | :------ |
| `«destructured»` | `Object` |
| › `envVar` | `string` |
| `message` | `string` |

#### Overrides

HealthCheckError.constructor

#### Defined in

[packages/decoupled-kit-health-check/src/classes/errors.ts:10](https://github.com/pantheon-systems/decoupled-kit-js/blob/5049fc03/packages/decoupled-kit-health-check/src/classes/errors.ts#L10)

## Properties

### cause

• `Optional` **cause**: `unknown`

#### Inherited from

HealthCheckError.cause

#### Defined in

node_modules/.pnpm/typescript@5.2.2/node_modules/typescript/lib/lib.es2022.error.d.ts:24

___

### message

• **message**: `string`

#### Inherited from

HealthCheckError.message

#### Defined in

node_modules/.pnpm/typescript@5.2.2/node_modules/typescript/lib/lib.es5.d.ts:1068

___

### name

• **name**: `string`

#### Inherited from

HealthCheckError.name

#### Defined in

node_modules/.pnpm/typescript@5.2.2/node_modules/typescript/lib/lib.es5.d.ts:1067

___

### stack

• `Optional` **stack**: `string`

#### Inherited from

HealthCheckError.stack

#### Defined in

node_modules/.pnpm/typescript@5.2.2/node_modules/typescript/lib/lib.es5.d.ts:1069

___

### prepareStackTrace

▪ `Static` `Optional` **prepareStackTrace**: (`err`: `Error`, `stackTraces`: `CallSite`[]) => `any`

#### Type declaration

▸ (`err`, `stackTraces`): `any`

Optional override for formatting stack traces

##### Parameters

| Name | Type |
| :------ | :------ |
| `err` | `Error` |
| `stackTraces` | `CallSite`[] |

##### Returns

`any`

**`See`**

https://v8.dev/docs/stack-trace-api#customizing-stack-traces

#### Inherited from

HealthCheckError.prepareStackTrace

#### Defined in

node_modules/.pnpm/@types+node@18.16.14/node_modules/@types/node/globals.d.ts:11

___

### stackTraceLimit

▪ `Static` **stackTraceLimit**: `number`

#### Inherited from

HealthCheckError.stackTraceLimit

#### Defined in

node_modules/.pnpm/@types+node@18.16.14/node_modules/@types/node/globals.d.ts:13

## Methods

### captureStackTrace

▸ `Static` **captureStackTrace**(`targetObject`, `constructorOpt?`): `void`

Create .stack property on a target object

#### Parameters

| Name | Type |
| :------ | :------ |
| `targetObject` | `object` |
| `constructorOpt?` | `Function` |

#### Returns

`void`

#### Inherited from

HealthCheckError.captureStackTrace

#### Defined in

node_modules/.pnpm/@types+node@18.16.14/node_modules/@types/node/globals.d.ts:4
