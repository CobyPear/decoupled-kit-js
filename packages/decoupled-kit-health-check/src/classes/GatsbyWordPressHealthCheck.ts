import dotenv from 'dotenv';
import { logger } from '../utils/logger';
import { resolveDotenvFile } from '../utils/resolveDotenvFile';
import {
	BackendNotSetError,
	DecoupledMenuError,
	InvalidCMSEndpointError,
	WPGatsbyPluginError,
} from './errors';
import { WordPressHealthCheck } from './HealthCheckBase';

export class GatsbyWordPressHealthCheck extends WordPressHealthCheck {
	endpoint: string;
	envVar: string;
	appUsername: string | undefined;
	#appPassword: string | undefined;
	log = logger;
	/**
	 * Resolves .env file if it exists and check for
	 * required environment variables, and sets the endpoint
	 * @param env - process.env
	 */
	constructor({ env }: { env: typeof process.env }) {
		super();
		if (process.env.NODE_ENV !== 'production') {
			dotenv.config({
				path: resolveDotenvFile(),
			});
		} else {
			console.log(
				'Production environment detected, skipping .env* resolution.',
			);
		}
		this.appUsername = env['WP_APPLICATION_USERNAME'];
		this.#appPassword = env['WP_APPLICATION_PASSWORD'];

		console.log('Checking for PANTHEON_CMS_ENDPOINT or WPGRAPHQL_URL...');
		const keys = ['WPGRAPHQL_URL', 'PANTHEON_CMS_ENDPOINT'];
		const backendVars: {
			[key in 'WPGRAPHQL_URL' | 'PANTHEON_CMS_ENDPOINT']?: string;
		} = {};
		Object.entries(env)
			.filter(([key]) => {
				if (keys.includes(key)) {
					return true;
				}
				return false;
			})
			.forEach(([key, value]) => {
				Object.assign(backendVars, { [key]: value });
			});

		const isVarSet = Object.keys(backendVars).length > 0;
		const areMultipleVarsSet = Object.keys(backendVars).length > 1;

		if (!isVarSet) {
			throw new BackendNotSetError({ envVar: 'WPGRAPHQL_URL' });
		} else {
			const setEndpoints = Object.keys(backendVars);
			this.log.success(
				`${setEndpoints.join(' and ')} ${
					areMultipleVarsSet ? 'are' : 'is'
				} set!`,
			);
			if (areMultipleVarsSet) {
				this.log.warn(
					`Both PANTHEON_CMS_ENDPOINT and WPGRAPHQL_URL are set.\n|____Using WPGRAPHQL_URL for remaining checks.`,
				);
			}
		}

		const [[envVar, endpoint]] = areMultipleVarsSet
			? Object.entries(backendVars).filter(([key]) => key === 'WPGRAPHQL_URL')
			: Object.entries(backendVars);

		const url = /^https:\/\//.test(endpoint) ? endpoint : `https://${endpoint}`;

		this.endpoint = url;
		this.envVar = envVar;
	}
	getURL() {
		const url = new URL(this.endpoint);
		url.pathname = '/wp/graphql';
		return url;
	}
	async checkFor200(url: URL) {
		try {
			const res = await fetch(url);
			const payload = (await res.json()) as {
				errors?: unknown[];
				data?: unknown;
			};
			if (payload.errors) {
				return false;
			} else if (payload.data) {
				return true;
			}
		} catch (error) {
			void error;
		}
		return false;
	}
	async validateEndpoint() {
		console.log('Validating CMS endpoint...');
		const cmsEndpoint = this.getURL();
		cmsEndpoint.searchParams.set('query', '{__typename}');
		const endpointIsValid = await this.checkFor200(cmsEndpoint);
		if (endpointIsValid) {
			this.log.success(`${this.envVar} is valid!`);
			return this;
		}
		throw new InvalidCMSEndpointError({
			endpointType: this.envVar,
			endpoint: this.getURL().host,
		});
	}
	async validateMenu() {
		console.log('Validating Menu Item endpoint...');
		const cmsEndpoint = this.getURL();
		const footerMenuQuery = /* graphql */ `query FooterMenuQuery {
			menu(idType: NAME, id: "Example Menu") {
				menuItems {
					edges {
						node {
							id
							path
							label
						}
					}
				}
			}
		}`;
		cmsEndpoint.searchParams.set('query', footerMenuQuery);
		const menuIsValid = await this.checkFor200(cmsEndpoint);
		if (menuIsValid) {
			this.log.success('Menu Items endpoint is valid!');
			return this;
		}
		throw new DecoupledMenuError({
			endpointType: this.envVar,
			endpoint: cmsEndpoint.host,
		});
	}
	async validateAuth() {
		console.log('Validating authentication...');
		const cmsEndpoint = this.getURL();

		if (!this.appUsername || !this.#appPassword) {
			this.log.warn('Auth not valid.');
			if (!this.appUsername) {
				this.log.warn('WP_APPLICATION_USERNAME is not set.');
				this.log.suggest(
					`Get the WP_APPLICATION_USERNAME here: ðŸ”— https://${cmsEndpoint.host}/wp/wp-admin/users.php`,
				);
			}
			if (!this.#appPassword) {
				this.log.warn('WP_APPLICATION_PASSWORD is not set.');
				this.log.suggest(
					`Set a new WP_APPLICATION_PASSWORD here by clicking edit: ðŸ”— https://${cmsEndpoint.host}/wp/wp-admin/users.php`,
				);
			}
			return this;
		}
		try {
			const credentials = `${String(this.appUsername)}:${String(
				this.#appPassword,
			)}`;
			const encodedCredentials = Buffer.from(credentials, 'binary').toString(
				'base64',
			);
			const query = /* graphql */ `query LatestPostsQuery {
				posts(where: { status: PRIVATE }) {
					edges {
						node {
							id
						}
					}
				}
			}`;
			cmsEndpoint.searchParams.set('query', query);
			const res = await fetch(cmsEndpoint, {
				headers: { Authorization: `Basic ${encodedCredentials}` },
			});
			const payload = (await res.json()) as {
				data: { posts: { edges: unknown[] } };
			};
			if (payload.data.posts.edges.length > 0) {
				this.log.success('Auth is valid!');
				return this;
			}
		} catch (error) {
			void error;
		}
		this.log.warn('Auth not valid.');
		this.log.suggest(
			'Ensure the WP_APPLICATION_USERNAME and WP_APPLICATION_PASSWORD are correct.',
		);
		return this;
	}
	async validateWPGatsbyPlugin() {
		const cmsEndpoint = this.getURL();
		const query = /* graphql */ `query checkWpGatsby {
			isWpGatsby
		}`;
		cmsEndpoint.searchParams.set('query', query);
		const res = await fetch(cmsEndpoint);
		const payload = (await res.json()) as {
			data?: { isWpGatsby: boolean };
			errors?: unknown[];
		};
		if (payload.errors) {
			throw new WPGatsbyPluginError({
				envVar: this.envVar,
				endpoint: cmsEndpoint.host,
			});
		} else if (payload?.data?.isWpGatsby) {
			this.log.success('WP Gatsby plugin is active!');
			return this;
		}
		return this;
	}
	// eslint-disable-next-line @typescript-eslint/require-await
	async validatePreview() {
		return this;
	}
}
