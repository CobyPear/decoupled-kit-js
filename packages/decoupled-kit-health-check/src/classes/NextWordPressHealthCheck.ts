import dotenv from 'dotenv';
import { logger } from '../utils/logger';
import { resolveDotenvFile } from '../utils/resolveDotenvFile';
import {
	BackendNotSetError,
	DecoupledMenuError,
	InvalidCMSEndpointError,
} from './errors';
import { WordPressHealthCheck } from './HealthCheckBase';

export class NextWordPressHealthCheck extends WordPressHealthCheck {
	endpoint: string;
	envVar: string;
	appUsername: string | undefined;
	#appPassword: string | undefined;
	#previewSecret: string | undefined;
	log = logger;
	#credentials: string | undefined = undefined;
	/**
	 * Resolves .env file if it exists and check for
	 * required environment variables, and sets the endpoint
	 * @param env - process.env
	 */
	constructor({ env }: { env: typeof process.env }) {
		super();
		this.appUsername = env['WP_APPLICATION_USERNAME'];
		this.#appPassword = env['WP_APPLICATION_PASSWORD'];
		this.#previewSecret = env['PREVIEW_SECRET'];

		console.log('Checking for PANTHEON_CMS_ENDPOINT or WPGRAPHQL_URL...');
		if (process.env.NODE_ENV !== 'production') {
			dotenv.config({
				path: resolveDotenvFile(),
			});
		} else {
			console.log(
				'Production environment detected, skipping .env* resolution.',
			);
		}
		const keys = ['WPGRAPHQL_URL', 'PANTHEON_CMS_ENDPOINT'];
		const backendVars: {
			[key in 'WPGRAPHQL_URL' | 'PANTHEON_CMS_ENDPOINT']?: string;
		} = {};
		Object.entries(env)
			.filter(([key]) => {
				if (keys.includes(key)) {
					return true;
				}
				return false;
			})
			.forEach(([key, value]) => {
				Object.assign(backendVars, { [key]: value });
			});

		const isVarSet = Object.keys(backendVars).length > 0;
		const areMultipleVarsSet = Object.keys(backendVars).length > 1;

		if (!isVarSet) {
			throw new BackendNotSetError('WPGRAPHQL_URL');
		} else {
			const setEndpoints = Object.keys(backendVars);
			this.log.success(
				`${setEndpoints.join(' and ')} ${
					areMultipleVarsSet ? 'are' : 'is'
				} set!`,
			);
			if (areMultipleVarsSet) {
				this.log.warn(
					`Both PANTHEON_CMS_ENDPOINT and WPGRAPHQL_URL are set.\n|____Using WPGRAPHQL_URL for remaining checks.`,
				);
			}
		}

		const [[envVar, endpoint]] = areMultipleVarsSet
			? Object.entries(backendVars).filter(([key]) => key === 'WPGRAPHQL_URL')
			: Object.entries(backendVars);

		let url;
		url = /^https:\/\//.test(endpoint) ? endpoint : `https://${endpoint}`;
		// ensure the url ends with /wp/graphql
		url = /\/wp\/graphql$/.test(url) ? url : `${url}/wp/graphql`;

		this.endpoint = url;
		this.envVar = envVar;
	}
	getURL() {
		return new URL(this.endpoint);
	}
	async checkFor200(url: URL) {
		try {
			const res = await fetch(url);
			const payload = (await res.json()) as {
				errors?: unknown[];
				data?: unknown;
			};
			if (payload.errors) {
				return false;
			} else if (payload.data) {
				return true;
			}
		} catch (error) {
			void error;
		}
		return false;
	}
	async validateEndpoint() {
		console.log('Validating CMS endpoint...');
		const cmsEndpoint = this.getURL();
		cmsEndpoint.searchParams.set('query', '{__typename}');
		const endpointIsValid = await this.checkFor200(cmsEndpoint);
		if (endpointIsValid) {
			this.log.success(`${this.envVar} is valid!`);
			return this;
		}
		throw new InvalidCMSEndpointError({
			endpointType: this.envVar,
			endpoint: this.getURL().host,
		});
	}
	async validateMenu() {
		console.log('Validating Menu Item endpoint...');
		const cmsEndpoint = this.getURL();
		const footerMenuQuery = /* graphql */ `query FooterMenuQuery {
			menu(idType: NAME, id: "Example Menu") {
				menuItems {
					edges {
						node {
							id
							path
							label
						}
					}
				}
			}
		}`;
		cmsEndpoint.searchParams.set('query', footerMenuQuery);
		const menuIsValid = await this.checkFor200(cmsEndpoint);
		if (menuIsValid) {
			this.log.success('Menu Items endpoint is valid!');
			return this;
		}
		throw new DecoupledMenuError({
			endpointType: this.envVar,
			endpoint: cmsEndpoint.host,
		});
	}
	async validateAuth() {
		console.log('Validating authentication...');
		const cmsEndpoint = this.getURL();
		if (!this.appUsername) {
			this.log.warn(
				'WP_APPLICATION_USERNAME is required for preview but is not set.',
			);
			this.log.suggest(
				`Get the WP_APPLICATION_USERNAME here: üîó https://${cmsEndpoint.host}/wp/wp-admin/users.php`,
			);
		}
		if (!this.#appPassword) {
			this.log.warn(
				'WP_APPLICATION_PASSWORD is required for preview but is not set.',
			);
			this.log.suggest(
				`Set a new WP_APPLICATION_PASSWORD here by clicking edit: üîó https://${cmsEndpoint.host}/wp/wp-admin/users.php`,
			);
		}
		try {
			const credentials = `${String(this.appUsername)}:${String(
				this.#appPassword,
			)}`;
			const encodedCredentials = Buffer.from(credentials, 'binary').toString(
				'base64',
			);
			const query = /* graphql */ `query LatestPostsQuery {
				posts(where: { status: PRIVATE }) {
					edges {
						node {
							id
						}
					}
				}
			}`;
			cmsEndpoint.searchParams.set('query', query);
			const res = await fetch(cmsEndpoint, {
				headers: { Authorization: `Basic ${encodedCredentials}` },
			});
			const payload = (await res.json()) as {
				data: { posts: { edges: unknown[] } };
			};
			if (payload.data.posts.edges.length > 0) {
				this.#credentials = encodedCredentials;
				this.log.success('Auth is valid!');
			} else {
				this.#credentials = '';
			}
		} catch (error) {
			void error;
		}
		this.log.warn('Auth not valid.');
		this.log.suggest(
			'Ensure the WP_APPLICATION_USERNAME and WP_APPLICATION_PASSWORD are correct.',
		);
		return this;
	}
	async validatePreview() {
		if (!this.#credentials) {
			console.log(
				'‚è≠  Skipping preview endpoint validation -- authorization required.',
			);
			return this;
		}
		console.log('Checking for PREVIEW_SECRET...');
		if (!this.#previewSecret) {
			this.log.warn('PREVIEW_SECRET env var is not set.');
			this.log.suggest(
				`To set a new secret, go to üîó https://${
					this.getURL().host
				}/admin/structure/dp-preview-site and edit the preview site you want to use.`,
			);
		} else {
			this.log.success('PREVIEW_SECRET is set.');
		}
		const cmsEndpoint = this.getURL();
		try {
			const query = /* graphql */ `query PostPreviewQuery {
				post(id: 1, idType: DATABASE_ID, asPreview: true) {
					title
					date
					featuredImage {
						node {
							altText
							sourceUrl
						}
					}
					content
				}
			}`;
			cmsEndpoint.searchParams.set('query', query);
			const res = await fetch(cmsEndpoint, {
				headers: {
					Authorization: `Basic ${this.#credentials}`,
				},
			});
			const payload = (await res.json()) as {
				errors?: unknown[];
				data?: { post: unknown };
			};
			if (payload?.data?.post) {
				this.log.success('Preview is valid!');
			}
		} catch (error) {
			void error;
		}
		this.log.warn('Could not fetch preview site.');
		return this;
	}
}
