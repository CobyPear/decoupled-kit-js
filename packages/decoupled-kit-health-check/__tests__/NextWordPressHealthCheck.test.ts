import { vi } from 'vitest';
import { NextWordPressHealthCheck } from '../src/classes/NextWordPressHealthCheck';

describe('NextWordPressHealthCheck', () => {
	beforeEach((context) => {
		vi.restoreAllMocks();
		context.logSpy = vi.spyOn(console, 'log');
		delete process.env['WPGRAPHQL_URL'];
		delete process.env['PANTHEON_CMS_ENDPOINT'];
		delete process.env['WP_APPLICATION_USERNAME'];
		delete process.env['WP_APPLICATION_SECRET'];
		delete process.env['PREVIEW_SECRET'];
	});
	it('should pass for a valid WPGRAPHQL_URL and invalid auth', async ({
		logSpy,
	}) => {
		process.env['WPGRAPHQL_URL'] = 'https://wordpress.test';
		const HC = new NextWordPressHealthCheck({ env: process.env });
		await HC.validateEndpoint()
			.then((hc) => hc.validateMenu())
			.then((hc) => hc.validateAuth())
			.then((hc) => hc.validatePreview())
			.catch(console.error);

		const result = logSpy.mock.calls.map(([call]) => call);
		expect(result).toMatchSnapshot();
	});
	it('should pass for a valid PANTHEON_CMS_ENDPOINT and invalid auth', async ({
		logSpy,
	}) => {
		process.env['PANTHEON_CMS_ENDPOINT'] = 'wordpress.test';
		const HC = new NextWordPressHealthCheck({ env: process.env });
		await HC.validateEndpoint()
			.then((hc) => hc.validateMenu())
			.then((hc) => hc.validateAuth())
			.then((hc) => hc.validatePreview())
			.catch(console.error);

		const result = logSpy.mock.calls.map(([call]) => call);
		expect(result).toMatchSnapshot();
	});
	it('should pass for a valid backend and valid auth but no preview secret', async ({
		logSpy,
	}) => {
		process.env['PANTHEON_CMS_ENDPOINT'] = 'wordpress.test';
		process.env['WP_APPLICATION_USERNAME'] = 'a1b2c3-d4e5g6';
		process.env['WP_APPLICATION_PASSWORD'] = 'mysecretsecret';

		const HC = new NextWordPressHealthCheck({ env: process.env });
		await HC.validateEndpoint()
			.then((hc) => hc.validateMenu())
			.then((hc) => hc.validateAuth())
			.then((hc) => hc.validatePreview())
			.catch(console.error);

		const result = logSpy.mock.calls.map(([call]) => call);
		expect(result).toMatchSnapshot();
	});
	it('should pass for a valid backend and valid auth', async ({ logSpy }) => {
		process.env['PANTHEON_CMS_ENDPOINT'] = 'wordpress.test';
		process.env['WP_APPLICATION_USERNAME'] = 'a1b2c3-d4e5g6';
		process.env['WP_APPLICATION_PASSWORD'] = 'mysecretsecret';
		process.env['PREVIEW_SECRET'] = 'testpreviewsecret!';

		const HC = new NextWordPressHealthCheck({ env: process.env });
		await HC.validateEndpoint()
			.then((hc) => hc.validateMenu())
			.then((hc) => hc.validateAuth())
			.then((hc) => hc.validatePreview())
			.catch(console.error);

		const result = logSpy.mock.calls.map(([call]) => call);
		expect(result).toMatchSnapshot();
	});
	// Unsure why this one is considered an error
	// and only works with `it.fails` but the following has different behavior.
	it.fails(
		'should show a helpful error message if no required env vars are set',
		async () => {
			const HC = new NextWordPressHealthCheck({ env: process.env });
			await HC.validateEndpoint()
				.then((hc) => hc.validateMenu())
				.then((hc) => hc.validateAuth())
				.then((hc) => hc.validatePreview())
				.catch((err) => console.log(err.message));
		},
	);
	it('should show a helpful error message if the backend is invalid', async ({
		logSpy,
	}) => {
		process.env['PANTHEON_CMS_ENDPOINT'] = 'invalid.wordpress.test';
		const HC = new NextWordPressHealthCheck({ env: process.env });
		await HC.validateEndpoint()
			.then((hc) => hc.validateMenu())
			.then((hc) => hc.validateAuth())
			.then((hc) => hc.validatePreview())
			.catch((err) => console.log(err.message));

		const result = logSpy.mock.calls.map(([call]) => call);
		expect(result).toMatchSnapshot();
	});
});
