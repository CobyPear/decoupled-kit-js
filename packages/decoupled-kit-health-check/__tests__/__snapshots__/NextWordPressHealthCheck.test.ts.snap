// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NextWordPressHealthCheck > should pass for a valid PANTHEON_CMS_ENDPOINT and invalid auth 1`] = `
[
  "Checking for PANTHEON_CMS_ENDPOINT or WPGRAPHQL_URL...",
  "No .env* file found, assuming production environment.",
  "|__‚úÖ PANTHEON_CMS_ENDPOINT is set!",
  "Validating CMS endpoint...",
  "|__‚úÖ PANTHEON_CMS_ENDPOINT is valid!",
  "Validating Menu Item endpoint...",
  "|__‚úÖ Menu Items endpoint is valid!",
  "Validating authentication...",
  "|__üí° WP_APPLICATION_USERNAME is required for preview but is not set.",
  "|____ Get the WP_APPLICATION_USERNAME here: üîó https://wordpress.test/wp/wp-admin/users.php",
  "|__üí° WP_APPLICATION_PASSWORD is required for preview but is not set.",
  "|____ Set a new WP_APPLICATION_PASSWORD here by clicking edit: üîó https://wordpress.test/wp/wp-admin/users.php",
  "|__üí° Auth not valid.",
  "|____ Ensure the WP_APPLICATION_USERNAME and WP_APPLICATION_PASSWORD are correct.",
  "‚è≠  Skipping preview endpoint validation -- authorization required.",
]
`;

exports[`NextWordPressHealthCheck > should pass for a valid WPGRAPHQL_URL and invalid auth 1`] = `
[
  "Checking for PANTHEON_CMS_ENDPOINT or WPGRAPHQL_URL...",
  "No .env* file found, assuming production environment.",
  "|__‚úÖ WPGRAPHQL_URL is set!",
  "Validating CMS endpoint...",
  "|__‚úÖ WPGRAPHQL_URL is valid!",
  "Validating Menu Item endpoint...",
  "|__‚úÖ Menu Items endpoint is valid!",
  "Validating authentication...",
  "|__üí° WP_APPLICATION_USERNAME is required for preview but is not set.",
  "|____ Get the WP_APPLICATION_USERNAME here: üîó https://wordpress.test/wp/wp-admin/users.php",
  "|__üí° WP_APPLICATION_PASSWORD is required for preview but is not set.",
  "|____ Set a new WP_APPLICATION_PASSWORD here by clicking edit: üîó https://wordpress.test/wp/wp-admin/users.php",
  "|__üí° Auth not valid.",
  "|____ Ensure the WP_APPLICATION_USERNAME and WP_APPLICATION_PASSWORD are correct.",
  "‚è≠  Skipping preview endpoint validation -- authorization required.",
]
`;

exports[`NextWordPressHealthCheck > should pass for a valid backend and valid auth 1`] = `
[
  "Checking for PANTHEON_CMS_ENDPOINT or WPGRAPHQL_URL...",
  "No .env* file found, assuming production environment.",
  "|__‚úÖ PANTHEON_CMS_ENDPOINT is set!",
  "Validating CMS endpoint...",
  "|__‚úÖ PANTHEON_CMS_ENDPOINT is valid!",
  "Validating Menu Item endpoint...",
  "|__‚úÖ Menu Items endpoint is valid!",
  "Validating authentication...",
  "|__‚úÖ Auth is valid!",
  "Checking for PREVIEW_SECRET...",
  "|__‚úÖ PREVIEW_SECRET is set.",
  "|__‚úÖ Preview is valid!",
]
`;

exports[`NextWordPressHealthCheck > should pass for a valid backend and valid auth but no preview secret 1`] = `
[
  "Checking for PANTHEON_CMS_ENDPOINT or WPGRAPHQL_URL...",
  "No .env* file found, assuming production environment.",
  "|__‚úÖ PANTHEON_CMS_ENDPOINT is set!",
  "Validating CMS endpoint...",
  "|__‚úÖ PANTHEON_CMS_ENDPOINT is valid!",
  "Validating Menu Item endpoint...",
  "|__‚úÖ Menu Items endpoint is valid!",
  "Validating authentication...",
  "|__‚úÖ Auth is valid!",
  "Checking for PREVIEW_SECRET...",
  "|__üí° PREVIEW_SECRET env var is not set.",
  "|____ To set a new secret, go to üîó https://wordpress.test/admin/structure/dp-preview-site and edit the preview site you want to use.",
  "‚è≠  Skipping preview endpoint validation -- PREVIEW_SECRET required.",
]
`;

exports[`NextWordPressHealthCheck > should show a helpful error message if the backend is invalid 1`] = `
[
  "Checking for PANTHEON_CMS_ENDPOINT or WPGRAPHQL_URL...",
  "No .env* file found, assuming production environment.",
  "|__‚úÖ PANTHEON_CMS_ENDPOINT is set!",
  "Validating CMS endpoint...",
  "|__‚ùå PANTHEON_CMS_ENDPOINT could not be fetched from.
Check that invalid.wordpress.test is valid and provides a 200 response",
]
`;
